### ISSUE TITLE
Unexpected behavior when using SNLE_A in PySBI

### ISSUE THREAD

**Issue by JunmiHu** [bgroenks96/SimulationBasedInference.jl#5]
## Description
The issue can be separated into two parts:

1. When calling solve with SNLE_A, it seems to only perform ensemble simulations once before training the network. My understanding of the algorithm is that sample simulations and network training should occur in an overarching for-loop. This is an implementation question/issue. 
2. Upon completion of sampling and training the network, the code seems to fail in building the posterior distribution, yielding an assertion error for a missing prior. This seems to be a bug. 

It seems to me that these are two separate issues, but they occur from the same example, so I have included it as a single issue here. 

## To reproduce
I have provided an example code below based on the linearode example to reproduce the issue.
```
using PythonCall
using SimulationBasedInference
using SimulationBasedInference.PySBI
using OrdinaryDiffEq
import Random

### Define the ODE inference problem
const rng = Random.MersenneTwister(1234);
ode_func(u,p,t) = -p[1]*u;
α_true = 0.2
ode_p = [α_true];
tspan = (0.0,10.0);
odeprob = ODEProblem(ode_func, [1.0], tspan, ode_p)

dt = 0.2
tsave = tspan[begin]+dt:dt:tspan[end];
n_obs = length(tsave);
observable = ODEObservable(:y, odeprob, tsave, samplerate=0.01);
forward_prob = SimulatorForwardProblem(odeprob, observable)

ode_solver = Tsit5()
forward_sol = solve(forward_prob, ode_solver);
true_obs = get_observable(forward_sol, :y)
noise_scale = 0.05
noisy_obs = true_obs .+ noise_scale*randn(rng, n_obs);

model_prior = prior(α=Beta(2,2));
noise_scale_prior = prior(σ=Exponential(0.1));

lik = IsotropicGaussianLikelihood(observable, noisy_obs, noise_scale_prior);
inference_prob = SimulatorInferenceProblem(forward_prob, ode_solver, model_prior, lik);

### Run inference
snpe_sol = solve(inference_prob, PySBI.SNLE_A(), rng=rng)
```
### Terminal output
This code yields the following command-line output:
```
WARNING: redefinition of constant Main.rng. This may fail, cause incorrect answers, or produce other errors.
[ Info: Running simulations
Running 1000 simulations.: 100%|██████████████████████████████████████████████| 1000/1000 [00:07<00:00, 126.70it/s]
[ Info: Training estimator
[ Info: Building posteriorEpochs trained: 369
ERROR: Python: AssertionError: You did not pass a prior. You have to pass the prior either at
            initialization `inference = SNLE(prior)` or to `.build_posterior
            (prior=prior)`.
Python stacktrace:
 [1] build_posterior
   @ sbi.inference.snle.snle_base ~/SBITesting.jl/.CondaPkg/env/lib/python3.11/site-packages/sbi/inference/snle/snle_base.py:328
Stacktrace:
  [1] pythrow()
    @ PythonCall.Core ~/.julia/packages/PythonCall/S5MOg/src/Core/err.jl:92
  [2] errcheck
    @ ~/.julia/packages/PythonCall/S5MOg/src/Core/err.jl:10 [inlined]
  [3] pycallargs
    @ ~/.julia/packages/PythonCall/S5MOg/src/Core/builtins.jl:212 [inlined]
  [4] pycall(f::Py, args::Py; kwargs::@Kwargs{})
    @ PythonCall.Core ~/.julia/packages/PythonCall/S5MOg/src/Core/builtins.jl:230
  [5] pycall
    @ ~/.julia/packages/PythonCall/S5MOg/src/Core/builtins.jl:220 [inlined]
  [6] Py
    @ ~/.julia/packages/PythonCall/S5MOg/src/Core/Py.jl:339 [inlined]
  [7] step!(solver::SimulationBasedInference.PySBI.PySBISolver{PySNE, SimulationBasedInference.PySBI.MCMCSampling})
    @ SimulationBasedInference.PySBI ~/.julia/packages/SimulationBasedInference/rr6Ff/ext/pysbi/pysne.jl:134
  [8] solve!(solver::SimulationBasedInference.PySBI.PySBISolver{PySNE, SimulationBasedInference.PySBI.MCMCSampling})
    @ SimulationBasedInference.PySBI ~/.julia/packages/SimulationBasedInference/rr6Ff/ext/pysbi/pysne.jl:144
  [9] #solve#1
    @ ~/.julia/packages/CommonSolve/JfpfI/src/CommonSolve.jl:23 [inlined]
 [10] top-level scope
    @ ~/SBITesting.jl/example/linearode/snle_test.jl:34
```

## Expected Behavior
1. According to the SBI python documentation (https://sbi-dev.github.io/sbi/tutorial/16_implemented_methods/), the SNLE API is implemented as such:
```
from sbi.inference import SNLE

inference = SNLE(prior)
proposal = prior
for _ in range(num_rounds):
    theta = proposal.sample((num_sims,))
    x = simulator(theta)
    _ = inference.append_simulations(theta, x).train()
    posterior = inference.build_posterior(mcmc_method="slice_np_vectorized",
                                          mcmc_parameters={"num_chains": 20,
                                                           "thin": 5})
    proposal = posterior.set_default_x(x_o)
```
Hence, I would expect this sampling-training to also be looped in SimulationBasedInference. I don't believe the current implementation is intended. 

2. I would expect the program to run without error.

It is worth noting that these issues seem to be SNLE_A specific, and that using SNPE_C runs without issue. 

## System info
OS: MacOS 14.4.1
julia version: 1.10.4

**Comment by bgroenks96** [comment]
Hi, thanks for raising the issue.

The prior not being passed to the algorithm is definitely an oversight and, if I read the `sbi` code correctly, is affecting SNPE as well. The SNPE algorithms default to using an improper prior if one is not supplied, so the issue was just more subtle.

As for the other issue, I did not intend to cover multi-round inference with the initial implementation. However, you're right that it would be more intuitive for the `solve` interface to do this since `init`/`step!` naturally imply an iterative process. I should be able to make this change with relatively minimal effort.

**Comment by bgroenks96** [comment]
Can you test with the latest commit on `main` and see what you get?

**Comment by JunmiHu** [comment]
Running the same code with multiple rounds now yields a different error:
"ERROR: Python: RuntimeError: number of categories cannot exceed 2^24". 

## To reproduce
```
using PythonCall
using SimulationBasedInference
using SimulationBasedInference.PySBI
using OrdinaryDiffEq
import Random

### Define the ODE inference problem
const rng = Random.MersenneTwister(1234);
ode_func(u,p,t) = -p[1]*u;
α_true = 0.2
ode_p = [α_true];
tspan = (0.0,10.0);
odeprob = ODEProblem(ode_func, [1.0], tspan, ode_p)

dt = 0.2
tsave = tspan[begin]+dt:dt:tspan[end];
n_obs = length(tsave);
observable = ODEObservable(:y, odeprob, tsave, samplerate=0.01);
forward_prob = SimulatorForwardProblem(odeprob, observable)

ode_solver = Tsit5()
forward_sol = solve(forward_prob, ode_solver);
true_obs = get_observable(forward_sol, :y)
noise_scale = 0.05
noisy_obs = true_obs .+ noise_scale*randn(rng, n_obs);

model_prior = prior(α=Beta(2,2));
noise_scale_prior = prior(σ=Exponential(0.1));

lik = IsotropicGaussianLikelihood(observable, noisy_obs, noise_scale_prior);
inference_prob = SimulatorInferenceProblem(forward_prob, ode_solver, model_prior, lik);

### Run inference
snpe_sol = solve(inference_prob, PySBI.SNLE_A(), rng=rng, num_rounds = 3)
```

## Stack trace
```
/SBITesting.jl/.CondaPkg/env/lib/python3.11/site-packages/sbi/utils/user_input_checks_utils.py:389: UserWarning: No prior bounds were passed, consider passing lower_bound
            and / or upper_bound if your prior has bounded support.
  warnings.warn(
[ Info: Starting iteration 1/3
[ Info: Running simulations
Running 1000 simulations.: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [00:08<00:00, 118.15it/s]
[ Info: Training estimator
[ Info: Building posteriorEpochs trained: 395
[ Info: Starting iteration 2/3
[ Info: Running simulations
ERROR: Python: RuntimeError: number of categories cannot exceed 2^24
Python stacktrace:
 [1] resample_given_potential_fn
   @ sbi.samplers.mcmc.init_strategy ~/sommerjobb2024/SBITesting.jl/.CondaPkg/env/lib/python3.11/site-packages/sbi/samplers/mcmc/init_strategy.py:112
 [2] <lambda>
   @ sbi.inference.posteriors.mcmc_posterior ~/sommerjobb2024/SBITesting.jl/.CondaPkg/env/lib/python3.11/site-packages/sbi/inference/posteriors/mcmc_posterior.py:355
 [3] <listcomp>
   @ sbi.inference.posteriors.mcmc_posterior ~/sommerjobb2024/SBITesting.jl/.CondaPkg/env/lib/python3.11/site-packages/sbi/inference/posteriors/mcmc_posterior.py:422
 [4] _get_initial_params
   @ sbi.inference.posteriors.mcmc_posterior ~/sommerjobb2024/SBITesting.jl/.CondaPkg/env/lib/python3.11/site-packages/sbi/inference/posteriors/mcmc_posterior.py:422
 [5] sample
   @ sbi.inference.posteriors.mcmc_posterior ~/sommerjobb2024/SBITesting.jl/.CondaPkg/env/lib/python3.11/site-packages/sbi/inference/posteriors/mcmc_posterior.py:275
 [6] simulate_for_sbi
   @ sbi.inference.base ~/sommerjobb2024/SBITesting.jl/.CondaPkg/env/lib/python3.11/site-packages/sbi/inference/base.py:493
Stacktrace:
  [1] pythrow()
    @ PythonCall.Core ~/.julia/packages/PythonCall/sQSpa/src/Core/err.jl:92
  [2] errcheck
    @ ~/.julia/packages/PythonCall/sQSpa/src/Core/err.jl:10 [inlined]
  [3] pycallargs
    @ ~/.julia/packages/PythonCall/sQSpa/src/Core/builtins.jl:222 [inlined]
  [4] pycall(f::Py, args::Py; kwargs::@Kwargs{proposal::Py, num_simulations::Int64, num_workers::Int64})
    @ PythonCall.Core ~/.julia/packages/PythonCall/sQSpa/src/Core/builtins.jl:237
  [5] pycall
    @ ~/.julia/packages/PythonCall/sQSpa/src/Core/builtins.jl:233 [inlined]
  [6] #_#11
    @ ~/.julia/packages/PythonCall/sQSpa/src/Core/Py.jl:357 [inlined]
  [7] step!(solver::SimulationBasedInference.PySBI.PySBISolver{PySNE, SimulationBasedInference.PySBI.MCMCSampling})
    @ SimulationBasedInference.PySBI ~/.julia/packages/SimulationBasedInference/Ieue9/ext/pysbi/pysne.jl:132
  [8] solve!(solver::SimulationBasedInference.PySBI.PySBISolver{PySNE, SimulationBasedInference.PySBI.MCMCSampling})
    @ SimulationBasedInference.PySBI ~/.julia/packages/SimulationBasedInference/Ieue9/ext/pysbi/pysne.jl:156
  [9] #solve#1
    @ ~/.julia/packages/CommonSolve/JfpfI/src/CommonSolve.jl:23 [inlined]
 [10] top-level scope
    @ ~/SBITesting.jl/example/linearode/snle_test.jl:34
```

**Comment by bgroenks96** [comment]
Very strange. That's either an issue with the simulation data format (most likely) or a bug in SBI. I'll take a look in a bit.

**Comment by bgroenks96** [comment]
Fixed by d0f1cffe4

Feel free to re-open if you encounter further issues.
